### PR TITLE
matplotlib recipe

### PR DESCRIPTION
**Pull request by tcaduser**
Resolves issue #712 and issue #857 

Successfully builds matplotlib and its dependency kiwisolver.  Leverages  existing recipes for numpy, freetype, etc.

Takes care of all known duplicate symbol issues.  Meets pep8 requirements.

To build:
```
python3 -mvenv venv
source venv/bin/activate
cd kivy-ios
pip install -e .
toolchain build matplotlib
```

screenshot when integrated with my application:

IMG_3256

### CODE REVIEW

**Comment by tcaduser** [comment]
FWIW, that screenshot is from my iphone, and not a simulator.

File: kivy_ios/tools/cpplink [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        continue
+
+    if opt.startswith('-arch'):
+        continue
+
+    if opt.startswith("-Wl,"):
+        continue
+
+    if opt.startswith("-W"):
+        continue
```
Comment (AndreMiras) [inline kivy_ios/tools/cpplink:93]:
```
opt.startswith(("-I", "-m", "-f", "-O", "-g", "-D", "-arch", "-Wl", "-W", "-stdlib=")):
    continue
```

File: kivy_ios/tools/cpplink [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+if not output.endswith('.so'):
+    result = subprocess.run([environ['CXX_ORIG'], *sys.argv[1:]])
+    sys.exit(result.returncode)
+
+
+libs = []
+objects = []
+
+i = 1
+while i < len(sys.argv):
```
Comment (AndreMiras) [inline kivy_ios/tools/cpplink:45]: Just curious, why not using the argparse module?
https://docs.python.org/3/library/argparse.html

Comment (AndreMiras) [reply kivy_ios/tools/cpplink:45]: Or maybe if we think argparse isn't suitable, make the code slightly more modular. That should help with readability, maintainability and testability. For instance:
```
def parse_arguments(args):
    output = None
    get_next = False
    libs = []
    objects = []
    i = 0
    while i < len(args):
        arg = args[i]
        if get_next:
            output = arg
            get_next = False
        elif arg.startswith('-o'):
            if arg == '-o':
                get_next = True
            else:
                output = arg[2:]
        elif arg.startswith("-l") or arg.startswith("-L"):
            libs.append(arg)
        elif arg in ("-r", "-pipe", "-no-cpp-precomp"):
            pass
        elif arg in ("--sysroot", "-isysroot", "-framework", "-undefined", "-macosx_version_min"):
            i += 1
        elif arg.startswith(("-I", "-m", "-f", "-O", "-g", "-D", "-arch", "-Wl", "-W", "-stdlib=")):
            pass
        elif arg.startswith("-"):
            logging.error(f"Unknown option: {arg}")
            sys.exit(1)
        elif arg.endswith('.o'):
            objects.append(arg)
        i += 1
    return output, libs, objects
```
That we could later use in a `main()` e.g.:
def main():
    # ...
    output, libs, objects = parse_arguments(sys.argv[1:])
    # ...
```
It just gives the idea, what do you think?

**Review by AndreMiras — COMMENTED** [review]
Awesome stuff, thanks for the PR and for addressing some of the comments and CI issues already.
Great progress overall.
CI is failing with:
```
2023-09-09T19:46:09.5904620Z [DEBUG   ] error: command '/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/kivy_ios-2023.8.24-py3.11.egg/kivy_ios/tools/cpplink' failed: No such file or directory
```
refs https://github.com/kivy/kivy-ios/actions/runs/6131086632/job/16643997732?pr=858
If I had to guess I would say `cpplink` should be added to the `MANIFEST.in`.

I've also gave some ideas to help with cpplink readability, let me know what you think

File: kivy_ios/tools/cpplink (hunk 2 above)
Comment (tcaduser) [reply kivy_ios/tools/cpplink:45]: Thanks for the point about cpplink being in the manifest, I have only been doing local installs.

The linking code was actually slightly modified from `tools/liblink`.  I will submit a slightly cleaned up version.

Comment (AndreMiras) [reply kivy_ios/tools/cpplink:45]: > The linking code was actually slightly modified from tools/liblink. I will submit a slightly cleaned up version.

Oh right, that code felt familiar

**Review by AndreMiras — APPROVED** [review]
LGTM, thanks for addressing.
Will merge after the build_updated_recipes build turns green

**Comment by tcaduser** [comment]
Looks like I needed to include some template files in the Manifest.  These were brought over from p4a.

**Comment by AndreMiras** [comment]
Great that you saw it too, thanks for addressing already.
Here's the error for records.
```
2023-09-09T22:05:49.5544410Z FileNotFoundError: [Errno 2] No such file or directory: '/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/kivy_ios-2023.8.24-py3.11.egg/kivy_ios/recipes/matplotlib/setup.cfg.template'
```
I'll be away from keyboard until tomorrow UTC to review/merge.
Good luck for the build :crossed_fingers:

**Comment by tcaduser** [comment]
I think I found the last issue in getting the right include file for numpy.

**Review by AndreMiras — APPROVED** [review]
CI happy well done :clap: 
Thanks for the quick follow up

**Comment by tcaduser** [comment]
Thanks so much for your help in getting this reviewed and accepted.

**Pull request merged**